lfsapi/auth.go

```diff
@@ -311,20 +314,41 @@ var (
 	authenticateHeaders = []string{"Lfs-Authenticate", "Www-Authenticate"}
 )
 
-func getAuthAccess(res *http.Response) creds.AccessMode {
-	for _, headerName := range authenticateHeaders {
-		for _, auth := range res.Header[headerName] {
-			pieces := strings.SplitN(strings.ToLower(auth), " ", 2)
-			if len(pieces) == 0 {
-				continue
+func getAuthAccess(res *http.Response, access creds.AccessMode, modes []creds.AccessMode) (creds.AccessMode, []creds.AccessMode) {
+	newModes := make([]creds.AccessMode, 0, len(modes))
+	for _, mode := range modes {
+		if access != mode {
+			newModes = append(newModes, mode)
+		}
+	}
+	if len(newModes) == 0 {
+		for _, mode := range creds.AllAccessModes() {
+			if mode != creds.NoneAccess {
+				newModes = append(newModes, mode)
 			}
+		}
+	}
```

Comment:
I confess I'm a little unclear on why this block is here ... I see, I think, that it causes the set of potential modes to be "reset", as it were, if none are left; that's why, in the accompanying test, the returned modes cycle through `negotiate`, `basic`, then `negotiate` again, `basic` again, and so on if one keeps going.  (I tried tinkering with this and it does cycle without end.)

What I'm unclear on is how this will interoperate with the [recursive call](https://github.com/git-lfs/git-lfs/blob/ee3088d4804bd2b699eb6a76aa1318d1d2283ed6/lfsapi/auth.go#L35) in `DoWithAuth()` above; see my comments in that function.

Reply:
Yeah, that was supposed to be intentional, but if you'd prefer we don't, I can remove it.